xcm/xcm-executor/src/lib.rs

```diff
@@ -242,17 +250,25 @@ impl<Config: config::Config> XcmExecutor<Config> {
 				ensure!(who.is_interior(), XcmError::EscalationOfPrivilege);
 				let mut origin = origin;
 				origin.append_with(who).map_err(|_| XcmError::MultiLocationFull)?;
-				let surplus =
-					Self::do_execute_xcm(origin, top_level, *message, weight_credit, None, trader)?;
+				let surplus = Self::do_execute_xcm(
+					origin,
+					top_level,
+					*message,
+					weight_credit,
+					None,
+					trader,
+					num_recursions + 1,
+				)?;
 				total_surplus = total_surplus.saturating_add(surplus);
 				None
 			},
 			_ => Err(XcmError::UnhandledXcmMessage)?, // Unhandled XCM message.
 		};
 
 		if let Some((mut holding, effects)) = maybe_holding_effects {
-			for effect in effects.into_iter() {
-				total_surplus += Self::execute_effects(&origin, &mut holding, effect, trader)?;
+			for effect in effects {
+				total_surplus +=
+					Self::execute_effects(&origin, &mut holding, effect, trader, num_recursions)?;
```

Comment:
```suggestion
					Self::execute_effects(&origin, &mut holding, effect, trader, num_recursions + 1)?;
```

This brings it (roughly) semantically in line with the decode depth limit.